### PR TITLE
AMDGPU/NewPM: Port SIFoldOperands to new pass manager

### PR DESCRIPTION
**Pull request by optimisan**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #include "R600.h"
 #include "R600TargetMachine.h"
 #include "SIFixSGPRCopies.h"
+#include "SIFoldOperands.h"
```
Comment (pravinjagtap) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:37]: Is this required ?

**Review by pravinjagtap — APPROVED** [review]
LGTM.

**Review by pravinjagtap — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIFoldOperands.cpp [hunk 2]
```diff
@@ -8,6 +8,7 @@
 //===----------------------------------------------------------------------===//
 //
 
+#include "SIFoldOperands.h"
 #include "AMDGPU.h"
 #include "GCNSubtarget.h"
 #include "MCTargetDesc/AMDGPUMCTargetDesc.h"
```
[the remark this answers is not in the document]
Comment (Akshat-Oke) [reply llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:14]: Seems like clang-format keeps the corresponding header include at the top

File: llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp (hunk 1 above)
Comment (Akshat-Oke) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:37]: Yes, it is used by the macros in `AMDGPUPassRegistry.def`.

File: llvm/lib/Target/AMDGPU/SIFoldOperands.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:14]: Yes the owning header should be first

File: llvm/lib/Target/AMDGPU/SIFoldOperands.cpp [hunk 3]
```diff
@@ -2246,3 +2254,15 @@ bool SIFoldOperands::runOnMachineFunction(MachineFunction &MF) {
...
+                                          MachineFunctionAnalysisManager &) {
+  SIFoldOperandsImpl Impl;
+  bool Changed = Impl.run(MF);
+  if (!Changed) {
+    return PreservedAnalyses::all();
+  }
+  auto PA = getMachineFunctionPassPreservedAnalyses();
+  PA.preserveSet<CFGAnalyses>();
+  return PA;
+}
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:2268]: Missing end of file line

File: llvm/lib/Target/AMDGPU/SIFoldOperands.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+namespace llvm {
+class SIFoldOperandsPass : public MachinePassInfoMixin<SIFoldOperandsPass> {
+public:
+  SIFoldOperandsPass() = default;
+  PreservedAnalyses run(MachineFunction &MF,
+                        MachineFunctionAnalysisManager &MFAM);
+};
+} // namespace llvm
+
+#endif
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIFoldOperands.h:23]: Missing end of file line

File: llvm/test/CodeGen/AMDGPU/flat-scratch-fold-fi.mir [hunk 5]
```diff
@@ -1,5 +1,6 @@
 # NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
 # RUN: llc -mtriple=amdgcn -mcpu=gfx900 -mattr=+enable-flat-scratch -run-pass=si-fold-operands -verify-machineinstrs -o - %s | FileCheck -check-prefix=GCN %s
+# RUN: llc -mtriple=amdgcn -mcpu=gfx900 -mattr=+enable-flat-scratch -passes=si-fold-operands -verify-machineinstrs -o - %s | FileCheck -check-prefix=GCN %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/flat-scratch-fold-fi.mir:3]: I don't want to double the number of run lines to test the pass itself. Just have a couple with the two forms? If you want to update all the tests, you can just use the new form

Comment (Akshat-Oke) [reply llvm/test/CodeGen/AMDGPU/flat-scratch-fold-fi.mir:3]: Alright, makes sense.

**Comment by Akshat-Oke** [comment]
Hold on, I can't find `MachinePassInfoMixin` now.

File: llvm/lib/Target/AMDGPU/SIFoldOperands.cpp [hunk 6]
```diff
@@ -121,11 +121,23 @@ class SIFoldOperands : public MachineFunctionPass {
...
+public:
+  static char ID;
+
+  SIFoldOperandsLegacy() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    if (skipFunction(MF.getFunction()))
+      return false;
+    SIFoldOperandsImpl Impl;
+    return Impl.run(MF);
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:139]:
```suggestion
    return SIFoldOperandsImpl().run(MF);
```

File: llvm/lib/Target/AMDGPU/SIFoldOperands.cpp [hunk 7]
```diff
@@ -2246,3 +2254,15 @@ bool SIFoldOperands::runOnMachineFunction(MachineFunction &MF) {
 
   return Changed;
 }
+
+PreservedAnalyses SIFoldOperandsPass::run(MachineFunction &MF,
+                                          MachineFunctionAnalysisManager &) {
+  SIFoldOperandsImpl Impl;
+  bool Changed = Impl.run(MF);
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:2261]:
```suggestion
  bool Changed = SIFoldOperandsImpl().run(MF);
```

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIFoldOperands.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+namespace llvm {
+class SIFoldOperandsPass : public PassInfoMixin<SIFoldOperandsPass> {
+public:
+  SIFoldOperandsPass() = default;
+  PreservedAnalyses run(MachineFunction &MF,
+                        MachineFunctionAnalysisManager &MFAM);
+};
+} // namespace llvm
+
+#endif
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIFoldOperands.h:23]:
```suggestion
#endif // LLVM_LIB_TARGET_AMDGPU_SIFOLDOPERANDS_H
```

**Review by arsenm — APPROVED** [review]

**Review by cdevadas — APPROVED** [review]

**Pull request merged**